asset/asset.go

```diff
@@ -801,20 +798,105 @@ var _ GenesisSigner = (*RawKeyGenesisSigner)(nil)
+			"version")
+	}
+
+	if newAsset.Genesis != initialGen && initialGen.Type != newAsset.Type {
+		return nil, fmt.Errorf("asset group type mismatch")
+	}
+
+	if scriptLeaf != nil && groupKey.TapscriptRoot == [32]byte{} {
+		return nil, fmt.Errorf("cannot script spend group key with " +
+			"empty tapscript root")
+	}
+
+	// Populate the sign descriptor with static and default fields.
+	genesisTweak := initialGen.ID()
+	signDesc := &lndclient.SignDescriptor{
+		KeyDesc:       groupKey.RawKey,
+		SingleTweak:   genesisTweak[:],
+		DoubleTweak:   nil,
+		TapTweak:      []byte{},
+		WitnessScript: []byte{},
+		HashType:      txscript.SigHashDefault,
+		InputIndex:    0,
+	}
+	if groupKey.TapscriptRoot != [32]byte{} {
+		signDesc.TapTweak = groupKey.TapscriptRoot[:]
+	}
+
+	tweakedGroupKey, err := GroupPubKey(
+		signDesc.KeyDesc.PubKey, signDesc.SingleTweak,
+		signDesc.TapTweak,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("cannot tweak group key: %w", err)
+	}
+
+	// Reimplement tapscript.InputAssetPrevOut to avoid a dependency cycle.
+	prevOutScript, err := txscript.NewScriptBuilder().
+		AddOp(txscript.OP_1).
+		AddData(schnorr.SerializePubKey(tweakedGroupKey)).
+		Script()
+	prevOut := &wire.TxOut{
+		Value:    int64(newAsset.Amount),
+		PkScript: prevOutScript,
+	}
+	prevAssets := commitment.InputSet{
+		ZeroPrevID: {newAsset},
+	}
 
-	groupPubKey, sig, err := genSigner.SignGenesis(
-		rawKey, initialGen, currentGen,
+	// Now, create the virtual transaction that represents this asset
+	// minting.
+	// TODO(jhb): resolve import cycles with tapscript package
+	virtualTx, _, err := tapscript.VirtualTx(newAsset, prevAssets)
+	if err != nil {
+		return nil, err
+	}
+	populatedVirtualTx := tapscript.VirtualTxWithInput(
+		virtualTx, newAsset, 0, nil,
+	)
+
+	// The last field to set for signing is the signing method needed, which
+	// can be inferred from the group key tapscript root and provided
+	// script leaf.
+	switch {
+	case bytes.Equal(signDesc.TapTweak, []byte{}):
+		signDesc.SignMethod = input.TaprootKeySpendBIP0086SignMethod
+	case scriptLeaf != nil:
+		signDesc.SignMethod = input.TaprootScriptSpendSignMethod
+		// TODO(jhb): extra validation here?
+		signDesc.WitnessScript = scriptLeaf.Script
+	case scriptLeaf == nil:
+		signDesc.SignMethod = input.TaprootKeySpendSignMethod
+	}
+
+	_, sig, err := genSigner.SignVirtualTx(
+		signDesc, populatedVirtualTx, prevOut,
 	)
 	if err != nil {
 		return nil, err
 	}
 
+	witness := wire.TxWitness{sig.Serialize()}
+	if signDesc.SignMethod == input.TaprootScriptSpendSignMethod {
+		witness = append(witness, signDesc.WitnessScript)
+		witness = append(witness, scriptLeaf.ControlBlock)
+	}
+
 	return &GroupKey{
-		RawKey:      rawKey,
-		GroupPubKey: *groupPubKey,
-		Witness:     wire.TxWitness{sig.Serialize()},
+		RawKey:        groupKey.RawKey,
+		GroupPubKey:   *tweakedGroupKey,
+		TapscriptRoot: groupKey.TapscriptRoot,
+		Witness:       witness,
```

Comment:
Why do we need to attach a witness here? The witness would reside in the asset being created (the very first one), then we just need to care about the key and tapscript root. Then this can be re-used everywhere, with the group minting identified based on the asset itself (prev id zero, etc). 

Or seems this is both the initial derivation _and_ also the first signing? 

For group keys in assets, we don't actually need to pass that witness along each time for serialization, so I assume this is just for book keeping or convenience purpose?

Reply:
In the minter, we set the group key info of the asset based on the output of `DeriveGroupKey`. I think ideally we'd modify `asset.New()` to copy the witness from here passed `GroupKey` into the genesis asset witness for the new asset.

Reply:
The actual group witness copy into the genesis asset witness is here in `asset.New()`:

https://github.com/lightninglabs/taproot-assets/blob/843584a12a7bcefd85003f584c24fe48e8705713/asset/asset.go#L951